clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -801,47 +887,71 @@ class SyclKernelFieldChecker
   bool IsInvalid = false;
   DiagnosticsEngine &Diag;
 
+  // Check whether the object should be disallowed from being copied to kernel.
+  // Return true if not copyable, false if copyable.
+  bool checkNotCopyableToKernel(const FieldDecl *FD, const QualType &FieldTy) {
+    if (FieldTy->isArrayType()) {
+      if (const auto *CAT = dyn_cast<ConstantArrayType>(FieldTy)) {
+        QualType ET = CAT->getElementType();
+        return checkNotCopyableToKernel(FD, ET);
+      } else
+        return Diag.Report(FD->getLocation(),
+                           diag::err_sycl_non_constant_array_type)
+               << FieldTy;
+    }
+
+    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams)
+      if (!FieldTy->isStandardLayoutType())
+        return Diag.Report(FD->getLocation(),
+                           diag::err_sycl_non_std_layout_type)
+               << FieldTy;
+
+    if (!FieldTy->isStructureOrClassType())
+      return false;
+
+    CXXRecordDecl *RD =
+        cast<CXXRecordDecl>(FieldTy->getAs<RecordType>()->getDecl());
+    if (!RD->hasTrivialCopyConstructor())
+      return Diag.Report(FD->getLocation(),
+                         diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+             << 0 << FieldTy;
+    if (!RD->hasTrivialDestructor())
+      return Diag.Report(FD->getLocation(),
+                         diag::err_sycl_non_trivially_copy_ctor_dtor_type)
+             << 1 << FieldTy;
+
+    return false;
+  }
+
 public:
   SyclKernelFieldChecker(Sema &S)
       : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
   bool isValid() { return !IsInvalid; }
 
-  void handleReferenceType(FieldDecl *FD, QualType FieldTy) final {
-    IsInvalid = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
-                << FieldTy;
-  }
-  void handleStructType(FieldDecl *FD, QualType FieldTy) final {
-    if (SemaRef.getASTContext().getLangOpts().SYCLStdLayoutKernelParams &&
-        !FieldTy->isStandardLayoutType())
-      IsInvalid =
-          Diag.Report(FD->getLocation(), diag::err_sycl_non_std_layout_type)
-          << FieldTy;
-    else {
-      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
-      if (!RD->hasTrivialCopyConstructor())
-
-        IsInvalid =
-            Diag.Report(FD->getLocation(),
-                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
-            << 0 << FieldTy;
-      else if (!RD->hasTrivialDestructor())
-        IsInvalid =
-            Diag.Report(FD->getLocation(),
-                        diag::err_sycl_non_trivially_copy_ctor_dtor_type)
-            << 1 << FieldTy;
-    }
+  bool handleReferenceType(FieldDecl *FD, QualType FieldTy) final {
+    bool bad = Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+               << FieldTy;
+    IsInvalid |= bad;
+    return !bad;
```

Comment:
AFAIK `bad` here is always true. @erichkeane , should we return `true` in handler functions to continue?

Reply:
`bad` is completely unnecessary here, I think 934 should just be: "IsInvalid = true", and return false.

You're right, Diag.Report ALWAYS returns true.   I'd also suggest having each of these functions just return isValid().